File: Scala/src/main/scala/TextbookTrack/Chapter03/BA3D.scala
package TextbookTrack.Chapter03

object BA3D {
  class DeBruijnGraph(text: String, k: Int) {
    import DeBruijnGraph.buildDeBruijnGraph

    def edges: Iterator[String] =
      for {
        (kMer, neighbours) <- adjacencyList.iterator
        neighbourString: String = neighbours.mkString(",")
      } yield s"$kMer -> $neighbourString"

    private val adjacencyList: Map[String, List[String]] = buildDeBruijnGraph(text, k)
  }

  object DeBruijnGraph {
    private def buildDeBruijnGraph(text: String, k: Int): Map[String, List[String]] =
      text
        .sliding(k)
        .map(kMer => (kMer.dropRight(1), kMer.drop(1)))
        .toList
        .groupBy{ case (prefix, _) => prefix }
        .view
        .mapValues(_.map{ case (_, suffix) => suffix })
        .toMap
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val k: Int = reader.next().toInt
    val text: String = reader.next()
    val graph = new DeBruijnGraph(text, k)
    graph.edges.foreach(println)
  }
}

File: Python/textbook_track/chapter03/ba3d.py
# Construct the De Bruijn Graph of a String
import sys

from collections import defaultdict


class DeBruijnGraph:
    def __init__(self, text, k):
        self._adjacency_list = build_de_bruijn_graph(text, k)

    def edges(self):
        for prefix, neighbours in self._adjacency_list.items():
            neighbour_string = ','.join(neighbours)
            yield f'{prefix} -> {neighbour_string}'


def build_de_bruijn_graph(text, k):
    d = defaultdict(list)
    for ix in range(len(text)-k+1):
        k_mer = text[ix:ix+k]
        prefix = k_mer[:-1]
        suffix = k_mer[1:]
        d[prefix].append(suffix)
    return dict(d)


def main():
    data = sys.stdin.read().splitlines()
    k = int(data[0])
    text = data[1]
    graph = DeBruijnGraph(text, k)
    for edge in graph.edges():
        print(edge)


if __name__ == '__main__':
    main()
